wgengine/router/router_linux.go

```diff
@@ -289,3 +326,261 @@ func (r *linuxRouter) restoreResolvConf() error {
+					return err
+				}
+				break
+			}
+		}
+
+		return nil
+	}
+
+	if err := del("filter", "INPUT"); err != nil {
+		return err
+	}
+	if err := del("filter", "FORWARD"); err != nil {
+		return err
+	}
+	if err := del("nat", "POSTROUTING"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// addBaseNetfilter4 installs the basic IPv4 netfilter framework for
+// Tailscale, in preparation for inserting more rules later.
+func (r *linuxRouter) addBaseNetfilter4() error {
+	// Create our own filtering chains, and hook them into the head of
+	// various main tables. If the hooks already exist, we don't try
+	// to fight for first place, because other software does the
+	// same. We're happy with "someplace up before most other stuff".
+	divert := func(table, chain string) error {
+		tsChain := "ts-" + strings.ToLower(chain)
+
+		chains, err := r.ipt4.ListChains(table)
+		if err != nil {
+			return err
+		}
+		found := false
+		for _, chain := range chains {
+			if chain == tsChain {
+				found = true
+				break
+			}
+		}
+		if found {
+			err = r.ipt4.ClearChain(table, tsChain)
+		} else {
+			err = r.ipt4.NewChain(table, tsChain)
+		}
+		if err != nil {
+			return err
+		}
+
+		args := []string{"-j", tsChain}
+		exists, err := r.ipt4.Exists(table, chain, args...)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return r.ipt4.Insert(table, chain, 1, args...)
+		}
+		return nil
+	}
+	if err := divert("filter", "INPUT"); err != nil {
+		return err
+	}
+	if err := divert("filter", "FORWARD"); err != nil {
+		return err
+	}
+	if err := divert("nat", "POSTROUTING"); err != nil {
+		return err
+	}
+
+	// Only allow CGNAT range traffic to come from tailscale0. There
+	// is an exception carved out for ranges used by ChromeOS, for
+	// which we fall out of the Tailscale chain.
+	//
+	// Note, this will definitely break nodes that end up using the
+	// CGNAT range for other purposes :(.
+	if err := r.ipt4.Append("filter", "ts-input", "!", "-i", r.tunname, "-s", "100.115.92.0/23", "-m", "comment", "--comment", "ChromeOS special ranges", "-j", "RETURN"); err != nil {
+		return err
+	}
+	if err := r.ipt4.Append("filter", "ts-input", "!", "-i", r.tunname, "-s", "100.64.0.0/10", "-j", "DROP"); err != nil {
+		return err
+	}
+
+	// Forward and masquerade packets that have the Tailscale subnet
+	// route bit set. The bit gets set by rules inserted into
+	// filter/FORWARD later on. We use packet marks here so both
+	// filter/FORWARD and nat/POSTROUTING can match on these packets
+	// of interest.
+	if err := r.ipt4.Append("filter", "ts-forward", "-m", "mark", "--mark", tailscaleSubnetRouteMark, "-j", "ACCEPT"); err != nil {
+		return err
+	}
+	if err := r.ipt4.Append("filter", "ts-forward", "-i", r.tunname, "-j", "DROP"); err != nil {
+		return err
+	}
+	// TODO(danderson): this should be optional.
+	if err := r.ipt4.Append("nat", "ts-postrouting", "-m", "mark", "--mark", tailscaleSubnetRouteMark, "-j", "MASQUERADE"); err != nil {
```

Comment:
To avoid unexpected behavior in strange edgecases, it might be advisable to be cautious and not perform MASQUERADING for traffic being sent 'into' the tailscale0 interface:
```suggestion
	if err := r.ipt4.Append("nat", "ts-postrouting", "!", "-o", r.tunname, "-m", "mark", "--mark", tailscaleSubnetRouteMark, "-j", "MASQUERADE"); err != nil {
```

Reply:
Thanks for taking a look!

In the marking logic, we only mark packets that ingressed on tailscaled0 and are destined for a subnet route. So, we should be fine here.

Do you have a scenario in mind that trips this edge case? If we're going to add this, I'd like to leave a comment detailing an example, so that in 6 months we're not confused by the apparent non-necessity.

Reply:
Correct, the marking logic should prevent this in most cases.

The (hypothetical and possibly unrealistic) case I had in mind is when we have a site-to-site setup with overlapping subnetworks announced at two locations.

I did run into this problem previously when building a network based on openvpn, and spent quite some time until I figured out that the packets were not actually dropped by my iptables rules but unexpectedly got a new source IP when routed into openvpn and that was why the traffic was not working as expected.

Let's assume we have relay node A being the HQ announcing the subnetwork 192.168.0.0/16.
Then branch office relay node B announcing 192.168.1.0/24.
And branch office relay node C announcing 192.168.2.0/24.
And C and B are unable to establish any connection between eachother (let's say due to some ongoing ISP peering issue).

Now a packet from 192.168.1.111 to 192.168.2.222 arrives at B, which sends it to A (as the 192.168.0.0/16 subnet points there).
At A, the packet is received and gets marked (as the 192.168.0.0/16 subnet route matches). Then it gets routed towards C (as on A, the 192.168.2.0/24 route exists and points there) and as it has the mark, it gets MASQUERADEd with the 100.64.X.Y IP of the tailscale0 interface.

Reply:
Ah, I see. Yes, this breaks in the case where we "hairpin" traffic through a different Tailscale node. However, currently we do not permit this, and it likely breaks because of the prefix accept lists we configure at the wireguard layer.

I have to think about this. Your suggestion is good defensive programming, but I'm worried about adding defenses against problems we don't have yet, because then future-me has to figure out why I did that :).

I also accidentally pushes this PR into master just now, so I'm going to close this pull request, but I'm still thinking about this change.